share/completions/meson.fish

```diff
@@ -1,52 +1,308 @@
 # Completions for the meson build system (http://mesonbuild.com/)
 
-set -l basic_arguments \
-    "h,help,show help message and exit" \
-    ",stdsplit,Split stdout and stderr in test logs" \
-    ",errorlogs,Print logs from failing test(s)" \
-    ",werror,Treat warnings as errors" \
-    ",strip,Strip targets on install" \
-    "v,version,Show version number and exit"
-
-set -l dir_arguments \
-    ",localedir,Locale data directory [share/locale]" \
-    ",sbindir,System executable directory [sbin]" \
-    ",infodir,Info page directory [share/info]" \
-    ",prefix,Installation prefix [/usr/local]" \
-    ",mandir,Manual page directory [share/man]" \
-    ",datadir,Data file directory [share]" \
-    ",bindir,Executable directory [bin]" \
-    ",sharedstatedir,Arch-agnostic data directory [com]" \
-    ",libdir,Library directory [system default]" \
-    ",localstatedir,Localstate data directory [var]" \
-    ",libexecdir,Library executable directory [libexec]" \
-    ",includedir,Header file directory [include]" \
-    ",sysconfdir,Sysconf data directory [etc]"
-
-for arg in $basic_arguments
-    set -l parts (string split , -- $arg)
-    if not string match -q "" -- $parts[1]
-        complete -c meson -s "$parts[1]" -l "$parts[2]" -d "$parts[3]"
+function __fish_meson_needs_command
+    set -l cmd (commandline -opc)
+    set -e cmd[1]
+    argparse -s 'v/version' -- $cmd 2>/dev/null
+    or return 0
+    not set -q argv[1]
+end
+
+function __fish_meson_using_command
+    set -l cmd (commandline -opc)
+    set -e cmd[1]
+    test (count $cmd) -eq 0
+    and return 1
+    contains -- $cmd[1] $argv
+    and return 0
+end
+
+function __fish_meson_builddir
+    # Consider the value of -C option to detect the build directory
+    set -l cmd (commandline -opc)
+    argparse -i 'C=' -- $cmd
+    if set -q _flag_C
+        echo $_flag_C
     else
-        complete -c meson -l "$parts[2]" -d "$parts[3]"
+        echo .
```

Comment:
If I remember correctly, meson doesn't support in-tree builds at all so this wouldn't work for commands that need the build directory and don't work without `-C` at all? I guess what I'm asking is are there any commands that work with `-C .` but don't work without `-C` at all? Because the inverse is true (e.g. `meson configure` can be run in a meson project directory without `-C` but fails with `-C .`).